### PR TITLE
1540218 slinky ground hx

### PR DESCRIPTION
**Pull request by Myoldmopar**
Making this a PR for convenience for conversation.

### CODE REVIEW

**Comment by DeadParrot** [comment]
@mmatts Good stuff but why the ```this->member``` usage? It is normally not needed within methods to disambiguate and just adds visual clutter. Where you have local variables with the same name as members it would be best to rename the locals to avoid silently hiding the members and confusing future developers.

**Comment by Myoldmopar** [comment]
That was my doing @deadparrot. Although some of the devs have an understanding of what an instance is, some don't. Putting the this pointer in there was my way of being clear when referencing a class instance variable or method. I'm sure the compiler doesn't care and creates the same assembly either way, so it shouldn't make any difference on performance. I won't take any hard stand on the issue, just keep in mind the status of the development team when discussing it.

**Comment by DeadParrot** [comment]
@Myoldmopar OK. I agree that it makes no difference to the compiled code: this issue is whether it helps users more than it hinders them. I can see some points in its favor. The factors I see are:
* Visual clutter that makes it harder to see what is going on (and could push lines to be wider than screen)
* Didn't have an analog of `this->` in Fortran so it shouldn't be needed as a transitional crutch
* It helps Intellisense work better so that is something
* Other than that good IDEs should know a variable is the local class member without `this->`
* If you refactor a function into a different class or to a non-class function (happens fairly often) you have to go remove all the old `this->` stuff.
* Using `this->` is probably more comfortable for Python-istas (but the self use in Python is considered a language flaw by many).
* If you have a naming ambiguity that requires `this->` then renaming the non-member variable is probably a good idea.
* I believe using `this->` when not needed is not as common but certainly there are those who like it: http://stackoverflow.com/questions/6779645/use-of-this-keyword-in-c

It would be a good idea to set up some standards before the code becomes a mish-mosh of different styles. C++ offers too much stylistic freedom to let everyone do their own thing.

**Comment by mmatts** [comment]
@Myoldmopar could you look at this briefly and let me know if I'm headed in a good direction WRT unit testing. I'm pretty much done implementing the g-function generation code and am planning to move on to GLHESlinky::calcGroundHeatExchanger tomorrow.

File: src/EnergyPlus/GroundHeatExchangers.cc [hunk 1]
```diff
@@ -163,52 +156,829 @@ namespace GroundHeatExchangers {
...
+		errFunc2 = std::erfc( 0.5 * sqrtDistDepth / sqrtAlphaT );
+	
+		return 4 * std::pow( Pi, 2.0 ) * errFunc1 / distance - errFunc2 / sqrtDistDepth;
+	};
+
+	//******************************************************************************
+
+	Real64
+	GLHESlinky::distance(
+	int const m,
```
Comment (Myoldmopar) [inline src/EnergyPlus/GroundHeatExchangers.cc]: Random note; some of these function signatures should have indented arguments.  Others already do.  Just noting.

File: src/EnergyPlus/GroundHeatExchangers.cc [hunk 2]
```diff
@@ -163,52 +156,829 @@ namespace GroundHeatExchangers {
...
+		// SUBROUTINE INFORMATION:
+		//       AUTHOR:          Matt Mitchell
+		//       DATE WRITTEN:    February, 2015
+		//       MODIFIED         na
+		//       RE-ENGINEERED    na
+
+		// PURPOSE OF THIS SUBROUTINE:
+		// Calculates the center-to-center distance between rings
+
+		// METHODOLOGY EMPLOYED:
```
Comment (Myoldmopar) [inline src/EnergyPlus/GroundHeatExchangers.cc]: Another random note: I'm less concerned about all these stubs.  I like having some information in the top of the function to help get author/written/methodology/purpose/references where applicable.  But feel free to clean out the bulk of these that aren't used.

**Comment by Myoldmopar** [comment]
@mmatts Holy shoot this is looking so good, it's exciting.  

With the performance PR merged in, you are now out of date with develop -- again -- sorry -- but it had to happen sometime.  Anyway, I can assist with resolving conflicts.  OK, I resolved them and pushed with beef0f7.  The conflicts were pretty minor so just take a quick pass at the updated ```Files Changed``` tab and make sure nothing looks goofy.

As for your comment, you were saying you were still passing indices into functions?  Could you give an example?

File: src/EnergyPlus/GroundHeatExchangers.cc [hunk 3]
```diff
@@ -163,52 +156,829 @@ namespace GroundHeatExchangers {
...
+		Real64 errFunc1;
+		Real64 errFunc2;
+		Real64 sqrtAlphaT;
+		Real64 sqrtDistDepth;
+		Real64 distance;
+
+		sqrtAlphaT = std::sqrt( diffusivityGround * t );
+		
+		distance = distToCenter( m, n, m1, n1 );
+		sqrtDistDepth = std::sqrt( std::pow( distance, 2.0 ) + 4 * std::pow( coilDepth, 2.0 ) );
```
Comment (DeadParrot) [inline src/EnergyPlus/GroundHeatExchangers.cc]: Performance note: pow_2( x ) is faster than std::pow( x, 2 ) and should be preferred. Yes, these really add up to matter in some of our profiles. We have pow_2 through pow_9 and a few others in ObjexxFCL/Fmath.hh.

Comment (Myoldmopar) [inline src/EnergyPlus/GroundHeatExchangers.cc]: Thanks @DeadParrot, your eagle eyes are very helpful!  I do have a performance sub-checklist on my review checklist, so I've added this as another thing to watch for in PRs.

**Comment by mmatts** [comment]
@Myoldmopar @DeadParrot thanks for the tips. I will get it cleaned up.

**Comment by mmatts** [comment]
@Myoldmopar Currently the arrays holding the (X,Y,Z) coordinates for each ring live [here](https://github.com/NREL/EnergyPlus/blob/1540218-SlinkyGroundHX/src/EnergyPlus/GroundHeatExchangers.hh#L196-198) at the object level. We're looping through all possible combinations of rings [here](https://github.com/NREL/EnergyPlus/blob/1540218-SlinkyGroundHX/src/EnergyPlus/GroundHeatExchangers.cc#L357-360) and then within that, to get the distances between any pair of rings, I just pass the correct indices [through](https://github.com/NREL/EnergyPlus/blob/1540218-SlinkyGroundHX/src/EnergyPlus/GroundHeatExchangers.cc#L386) [several](https://github.com/NREL/EnergyPlus/blob/1540218-SlinkyGroundHX/src/EnergyPlus/GroundHeatExchangers.cc#L922) [functions](https://github.com/NREL/EnergyPlus/blob/1540218-SlinkyGroundHX/src/EnergyPlus/GroundHeatExchangers.cc#L852) to the [distance function](https://github.com/NREL/EnergyPlus/blob/1540218-SlinkyGroundHX/src/EnergyPlus/GroundHeatExchangers.cc#L490). Would it be easier for unit testing purposes to just pass the (X,Y,Z) center values for each pair of rings through, or can we make this work.

**Comment by Myoldmopar** [comment]
@mmatts I noticed clang is still failing. Are you by any chance doing something like array().name() where you are trying to retrieve all the 'name' variables on each member of the array? I'm on my phone so I can't tell right now, but I'll look into it later.

**Comment by mmatts** [comment]
@Myoldmopar is [this](https://github.com/NREL/EnergyPlus/blob/1540218-SlinkyGroundHX/src/EnergyPlus/GroundHeatExchangers.cc#L1381) something along the lines of what you're talking about?

**Comment by Myoldmopar** [comment]
Yes, that is what I'm talking about. Vertical should already have that implemented. Any new structures won't, and really shouldn't. Now, in not sure why it wouldn't be failing to build on your local machine...

**Comment by DeadParrot** [comment]
@Myoldmopar to clarify how the "member array" hack works in ObjexxFCL, because name() is already set up as a member for ''all'' 1D arrays `verticalGLHE.name()` should work fine without any customization required specific to `verticalGLHE`, as long as its elements have a `name` std::string member.

We may want to discourage new use of member array syntax but for common usage such as `name()` it should work "out of the box". When someone wants to expose a member with a name not already set up then it would require the (discouraged) addition to the appropriate *.Projects.MArray.hh file.

**Comment by Myoldmopar** [comment]
Looks like it might still be having trouble with Name. I will look tomorrow morning, but it might be that the .Name() is still being used inside the sim routine. Maybe. Just guessing while on my phone...

**Comment by Myoldmopar** [comment]
@mmatts OK, I got it building with commit beef740.  I also cleanup up some unneeded trailing semicolons.  I didn't actually run it, but it should at least build now. It is very interesting that Windows seemed to build that just fine.

One thing to note is that in the Sim function, we won't be able to do the string comparison to check which type we have; it will be too slow.  I think it's probably time to bite the bullet and store pointers to the base class in a single array rather than two separate arrays, one per type.  Let's chat today about it.

**Comment by Myoldmopar** [comment]
@mmatts the test results look good, just tiny diff's in a few files, I'm guessing similar to what you were seeing locally. I think this is ready to merge, let me know if you were all doing anything else.

**Comment by mmatts** [comment]
@Myoldmopar let's hold off for a little bit. I just tracked down the source of the tiny diffs. I'm going to push that right now. Pending CI and any recommendations from you or others on the unit tests, it should be good.

**Comment by Myoldmopar** [comment]
@mmatts Yay!  100% passed!   :+1: 

I'll let CI make a pass at 10d0493 and this should drop in after.  Great work.

**Comment by Myoldmopar** [comment]
Oh, @mmatts the CI results might turn red due to diffs since develop has moved since your commit.  I'll look over the results and deem whether we need to merge develop or not.

**Pull request merged**